enhancements/olm-deprecate-copied-csvs.md

```diff
@@ -0,0 +1,207 @@
+  workloads during upgrades. Ensure the components leverage best practices in handling [voluntary disruption](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/). Any exception to this should be
+  identified and discussed here.
+- Micro version upgrades - users should be able to skip forward versions within a
+  minor release stream without being required to pass through intermediate
+  versions - i.e. `x.y.N->x.y.N+2` should work without requiring `x.y.N->x.y.N+1`
+  as an intermediate step.
+- Minor version upgrades - you only need to support `x.N->x.N+1` upgrade
+  steps. So, for example, it is acceptable to require a user running 4.3 to
+  upgrade to 4.5 with a `4.3->4.4` step followed by a `4.4->4.5` step.
+- While an upgrade is in progress, new component versions should
+  continue to operate correctly in concert with older component
+  versions (aka "version skew"). For example, if a node is down, and
+  an operator is rolling out a daemonset, the old and new daemonset
+  pods must continue to work correctly even while the cluster remains
+  in this partially upgraded state for some time.
+
+## Implementation History
+
+- N/A
+
+## Drawbacks
+
+- The Operator API is not a "drop-in-place" for Copied CSVs
+- There is no equivalent to namespace scoped CSV `List` requests for `Operator` resources; i.e. users must perform individual `Get` requests for each `Operator` they want to inspect, and must comb through `Operator` status to determine if an operator can act on a target namespace.
+
+## Alternatives
+
+### Command-line Option Only
+
+Update the olm-operator `Deployment` to toggle the Copied CSV feature.
+
+#### Cons
+
+On an OpenShift cluster, [a CVO override](https://docs.openshift.com/container-platform/4.8/logging/config/cluster-logging-maintenance-support.html#unmanaged-operators_cluster-logging-unsupported) would be required to prevent the updated command-line option in the olm-operator `Deployment` from reverting, but this also puts clusters into an unsupported state and prevents cluster upgrades.
+
+### OpenShift FeatureGates
+
+Define a new OpenShift [FeatureGate](https://github.com/openshift/api/blob/master/config/v1/types_feature.go) feature set that, when disabled, disables Copied CSVs. This would be preferred over any command-line option and would require no CVO override.
+
+#### Cons
+
+- Requires OLM to be OpenShift-aware
+  - Could be addressed with a downstream-only controller that deploys OLM with the correct options
+- Requires changes OpenShift (a new feature set)
+
+### OLM Config CRD
+
+Add a new CRD to define OLM-specific configuration instead of using the `olm-config` `ConfigMap`, adding more type-safety.
+
+#### Cons
+
+- Adds a new API to maintain
+- Adding APIs is questionably backportable (for OpenShift releases)
+- Singleton CRs beg for singleton enforcement, which today requires validating webhooks, increasing implementation complexity
```

Comment:
good to know!

> , if you create another one w/ a different name, it's ignored.

this is one of my major UX gripes around singletons -- it would be really nice if we could codify singletons in our schema definitions directly.

Reply:
From a purist perspective, I agree with Nick that singleton-style objects can be annoying to debug as a user if the cluster doesn't do anything to warn you that you messed something up.

If we go the singleton CRD route, we _could_ watch all objects of this type and add a condition that signals whether the configuration has been applied. That way, there's a signal to the creator about what we did (or didn't do) with it.

Reply:
Fair point.